File: packages/semantics/src/semantics-block/suggest.ts
// Nearest-match over a closed vocabulary (Levenshtein ≤ maxDistance). Used by the
// TTR-SEM-200/201/202 diagnostics to offer "did you mean …" on an unknown key /
// role / kind. Kept local to the semantics-block module — no TS-side edit-distance
// helper existed to reuse (the lint "suggestion" fixes are a different concept).

/** Levenshtein edit distance between `a` and `b` (classic DP). */
export function editDistance(a: string, b: string): number {
  const m = a.length;
  const n = b.length;
  if (m === 0) return n;
  if (n === 0) return m;
  let prev = Array.from({ length: n + 1 }, (_, j) => j);
  let curr = new Array<number>(n + 1);
  for (let i = 1; i <= m; i++) {
    curr[0] = i;
    for (let j = 1; j <= n; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      curr[j] = Math.min(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + cost);
    }
    [prev, curr] = [curr, prev];
  }
  return prev[n];
}

/**
 * The single closest candidate within `maxDistance` (default 2), or undefined if
 * none is close enough. Ties break to the first candidate in `candidates` order
 * (the vocabulary's declaration order), keeping suggestions deterministic.
 */
export function nearestMatch(
  input: string,
  candidates: ReadonlyArray<string>,
  maxDistance = 2,
): string | undefined {
  let best: string | undefined;
  let bestDist = maxDistance + 1;
  for (const c of candidates) {
    const d = editDistance(input, c);
    if (d < bestDist) {
      best = c;
      bestDist = d;
    }
  }
  return bestDist <= maxDistance ? best : undefined;
}

File: packages/kotlin/ttr-semantics/src/main/kotlin/org/tatrman/ttr/semantics/semanticsblock/Suggest.kt
// SPDX-License-Identifier: Apache-2.0
package org.tatrman.ttr.semantics.semanticsblock

/**
 * Nearest-match over a closed vocabulary (Levenshtein ≤ maxDistance). Used by the
 * TTR-SEM-200/201/202 diagnostics to offer "did you mean …" on an unknown key /
 * role / kind. Mirrors `semantics-block/suggest.ts`.
 */
object Suggest {
    /** Levenshtein edit distance between [a] and [b] (classic DP). */
    fun editDistance(
        a: String,
        b: String,
    ): Int {
        val m = a.length
        val n = b.length
        if (m == 0) return n
        if (n == 0) return m
        var prev = IntArray(n + 1) { it }
        var curr = IntArray(n + 1)
        for (i in 1..m) {
            curr[0] = i
            for (j in 1..n) {
                val cost = if (a[i - 1] == b[j - 1]) 0 else 1
                curr[j] = minOf(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + cost)
            }
            val tmp = prev
            prev = curr
            curr = tmp
        }
        return prev[n]
    }

    /**
     * The single closest candidate within [maxDistance] (default 2), or null if
     * none is close enough. Ties break to the first candidate in declaration order,
     * keeping suggestions deterministic.
     */
    fun nearestMatch(
        input: String,
        candidates: List<String>,
        maxDistance: Int = 2,
    ): String? {
        var best: String? = null
        var bestDist = maxDistance + 1
        for (c in candidates) {
            val d = editDistance(input, c)
            if (d < bestDist) {
                best = c
                bestDist = d
            }
        }
        return if (bestDist <= maxDistance) best else null
    }
}
